File: app/src/main/java/com/example/weizhenbin/wangebug/image/ImageLoader.java
package com.example.weizhenbin.wangebug.image;

import android.content.Context;
import android.widget.ImageView;

/**
 * Created by weizhenbin on 2018/8/15.
 */

public class ImageLoader implements IImageLoader {


    private IImageLoader realImageLoader;

    private static ImageLoader imageLoader;

    private ImageLoader(){

    }

    public void init(IImageLoader iImageLoader){
        if(iImageLoader==null){
            throw new NullPointerException("iImageLoader 不能为空");
        }
        this.realImageLoader=iImageLoader;
    }


    public static ImageLoader getImageLoader(){
        if(imageLoader==null){
            synchronized (ImageLoader.class){
                if (imageLoader==null){
                    imageLoader=new ImageLoader();
                }
            }
        }
        return imageLoader;
    }


    @Override
    public void imageLoader(Context context, ImageView imageView, String url) {
        checkNull();
        realImageLoader.imageLoader(context,imageView,url);
    }

    @Override
    public void imageLoader(Context context, ImageView imageView, String url, ImageConfig imageConfig) {
        checkNull();
        realImageLoader.imageLoader(context,imageView,url,imageConfig);
    }

    @Override
    public void imageLoader(Context context, ImageView imageView, String url, IImageLoadListener iImageLoadListener) {
        checkNull();
        realImageLoader.imageLoader(context,imageView,url,iImageLoadListener);
    }

    @Override
    public void imageLoader(Context context, ImageView imageView, String url, ImageConfig imageConfig, IImageLoadListener iImageLoadListener) {
        checkNull();
        realImageLoader.imageLoader(context,imageView,url,imageConfig,iImageLoadListener);
    }

    @Override
    public void loadBitmap(Context context, String url, IImageLoadListener iImageLoadListener) {
        checkNull();
        realImageLoader.loadBitmap(context,url,iImageLoadListener);
    }

    private void checkNull(){
        if (realImageLoader==null){
            throw new NullPointerException("没有调用init");
        }
    }

}

File: app/src/main/java/com/example/weizhenbin/wangebug/image/ImageLoader.kt
package com.example.weizhenbin.wangebug.image

import android.content.Context
import android.widget.ImageView

/**
 * Created by weizhenbin on 2018/8/15.
 */

class ImageLoader private constructor() : IImageLoader {


    private var realImageLoader: IImageLoader? = null

    fun init(iImageLoader: IImageLoader?) {
        if (iImageLoader == null) {
            throw NullPointerException("iImageLoader 不能为空")
        }
        this.realImageLoader = iImageLoader
    }


    override fun imageLoader(context: Context, imageView: ImageView, url: String) {
        checkNull()
        realImageLoader!!.imageLoader(context, imageView, url)
    }

    override fun imageLoader(context: Context, imageView: ImageView, url: String, imageConfig: ImageConfig?) {
        checkNull()
        realImageLoader!!.imageLoader(context, imageView, url, imageConfig)
    }

    override fun imageLoader(context: Context, imageView: ImageView, url: String, iImageLoadListener: IImageLoadListener?) {
        checkNull()
        realImageLoader!!.imageLoader(context, imageView, url, iImageLoadListener)
    }

    override fun imageLoader(context: Context, imageView: ImageView, url: String, imageConfig: ImageConfig?, iImageLoadListener: IImageLoadListener?) {
        checkNull()
        realImageLoader!!.imageLoader(context, imageView, url, imageConfig, iImageLoadListener)
    }

    override fun loadBitmap(context: Context, url: String, iImageLoadListener: IImageLoadListener?) {
        checkNull()
        realImageLoader!!.loadBitmap(context, url, iImageLoadListener)
    }

    private fun checkNull() {
        if (realImageLoader == null) {
            throw NullPointerException("没有调用init")
        }
    }


    private object SingletonHolder{
        val holder=ImageLoader()
    }



    companion object {

        fun getImageLoader()=SingletonHolder.holder
    }




}
